src/vision/status_quo/alan_builds_a_task_scheduler.md

```diff
@@ -0,0 +1,704 @@
+
+    pub fn trigger(&self, id: u64, x: String) -> io::Result<()> {
+        let mut _mu;
+        match self.requests.try_write() {
+            Some(guard) => _mu = guard,
+            None => return Err(...),
+        }
+
+        let request_tx;
+        match _mu.get(&id) {
+            Some(ch) => request_tx = ch,
+            None => return Err(...),
+        }
+
+        match request_tx.send(x) {
+            Ok(_) => _mu.remove(&id),
+            Err(e) => return Err(...),
+        }
+
+        Ok(())
+    }
+}
+```
+
+Alan is still not sure if `crossbeam_channel` is safe for async programming and whether he should instead use another crate [`async_std::channel`](https://docs.rs/async-std/1.9.0/async_std/channel/index.html). While `crossbeam_channel` seems to work, Alan is not confident about his choice. Disgruntled with seemingly unnecessary divergence in the community, Alan wonders why all those cool improvements had not been made back to Rust core std libraries.
+
+### **Implementing task applier**
+
+Alan implements a task applier, which simply echoes the requested message, as in Go:
+
+```go
+type EchoManager interface {
+	apply(req *EchoRequest) (string, error)
+}
+
+type echoManager struct {
+	mu sync.RWMutex
+}
+
+func (ea *echoManager) apply(req *EchoRequest) (string, error) {
+	ea.mu.Lock()
+	defer ea.mu.Unlock()
+	switch req.Kind {
+	case "create":
+		return fmt.Sprintf("SUCCESS create %q", req.Message), nil
+	case "delete":
+		return fmt.Sprintf("SUCCESS delete %q", req.Message), nil
+	default:
+		return "", fmt.Errorf("unknown request %q", req)
+	}
+}
+```
+
+Having implemented event notifier above, Alan is now somewhat familiar with Rust mutex and writes the following Rust code:
+
+```rust
+// 1st version
+use async_std::sync::RwLock;
+
+pub struct Manager {
+    mu: RwLock<()>,
+}
+
+impl Manager {
+    pub fn new() -> Self {
+        Self {
+            mu: RwLock::new(()),
+        }
+    }
+
+    pub fn apply(&self, req: &Request) -> io::Result<String> {
+        let _mu;
+        match self.mu.try_write() {
+            Some(guard) => _mu = guard,
+            None => return Err(...),
+        }
+        match req.kind.as_str() {
+            "create" => Ok(format!(
+                "SUCCESS create {}",
+                to_string(req.message.to_owned())
+            )),
+            "delete" => Ok(format!(
+                "SUCCESS delete {}",
+                to_string(req.message.to_owned())
+            )),
+            _ => Err(...),
+        }
+    }
+}
+```
+
+The code compiles and thus must be safe. However, after discussing the code with [Barbara][], Alan learns that `std::sync::Mutex` protects data from concurrent access, but `std::sync::Mutex` itselt must be also protected between threads. And this is where [`std::sync::Arc`](https://doc.rust-lang.org/std/sync/struct.Arc.html) comes in to provide safe multi-threaded access to the `Mutex`. Deeply confused, Alan made a quick fix to wrap `Mutex` with `Arc`:
```

Comment:
> The code compiles and thus must be safe. However, after discussing the code with [Barbara][], Alan learns that `std::sync::Mutex` protects data from concurrent access, but `std::sync::Mutex` itselt must be also protected between threads.

Hmm this isn't really true, or at least it's phrased in a potentially confusing way. `Arc` must be used to make the type useable from several threads. The code will not compile if you try to use it from multiple threads. This is relevant, but perhaps there was a misunderstanding in the communication between Barbara and Alan?